sympy/core/relational.py

```diff
@@ -1138,3 +1032,265 @@ def _eval_relation(cls, lhs, rhs):
     '<': StrictLessThan,
     'lt': StrictLessThan,
 }
+
+
+def _n2(a, b):
+    """Return (a - b).evalf(2) if a and b are comparable, else None.
+    This should only be used when a and b are already sympified.
+    """
+    # /!\ it is very important (see issue 8245) not to
+    # use a re-evaluated number in the calculation of dif
+    if a.is_comparable and b.is_comparable:
+        dif = (a - b).evalf(2)
+        if dif.is_comparable:
+            return dif
+
+
+def fuzzy_resolve(lhs, rhs, op, cls):
+    """
+    Resolves op to S.true if cmp returns true.
+    or to S.false if cmp returns false
+    or to an unevaluated if cmp returns none.
+    """
+    val = op(lhs, rhs)
+    if val is None:
+        return cls(lhs, rhs, evaluate=False)
+    elif isinstance(val,bool):
+        if val is True:
+            return S.true
+        else:
+            return S.false
+    else:
+        return val
+
+
+def is_lt(lhs, rhs):
+    retval = None
+    if lhs.is_extended_real and rhs.is_extended_real and hasattr(lhs, "_eval_is_ge"):
+        retval = fuzzy_not(getattr(lhs, "_eval_is_ge")(rhs))
+    elif hasattr(lhs, "_eval_is_lt"):
+        retval = getattr(lhs, "_eval_is_lt")(rhs)
+    if retval is None:
+        retval = cmp(lhs, rhs, "<")
+    return retval
+
+
+def is_le(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_is_gt"):
+        retval = fuzzy_not(getattr(lhs, "_eval_is_gt")(rhs))
+    if retval is None:
+        retval = cmp(lhs, rhs, "<=")
+    return retval
+
+
+def is_gt(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_is_gt"):
+        retval = getattr(lhs, "_eval_is_gt")(rhs)
+    if retval is None:
+        retval = cmp(lhs, rhs, ">")
+    return retval
+
+
+def is_ge(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_is_ge"):
+        retval = getattr(lhs, "_eval_is_ge")(rhs)
+    if retval is None:
+        retval = cmp(lhs, rhs, ">=")
+    return retval
+
+
+def is_eq(lhs, rhs):
+    retval = None
+    if hasattr(lhs, "_eval_Eq"):
+        retval = getattr(lhs, "_eval_Eq")(rhs)
```

Comment:
Why is this using `getattr`? Can it not just be:
```
retval = lhs._eval_Eq(rhs)
```

Reply:
What if _eval_Eq doesn't exist on lhs?

Reply:
Sometimes _eval_Eq doesn't exist on the object. And since we're not really using inheritance for this, if i try to call _eval_Eq on an object that does not exist it will throw an error.

Reply:
I mean specifically this line. If we've already checked with `hasattr` there's no need to use `getattr`. The two arg form of `getattr` raises if the attribute is not found anyway.

We could just add the `_eval_Eq` method on `Basic` and have it return `None`.

I think it's fine to document that these functions are internal and only for `Basic` and that they don't sympify.

Reply:
oh, just looked up how hasattr works, I see what you're saying now. And actually implementing _eval_is_X on Basic objects, really does make sense to me. Since everything will be sympified by this point. So, if there's no problem with that, that will be the approach I will take.